Implemented jwt auth

SedovMykyta opened the pull request:
feat - Implemented jwt auth;
feat - Added auth service and controller;
feat - Added 2 new custom exception;
feat - Added new methods in user service, controller; 
fix - Fixed create, update methods in user service;

Commit:
feat - Implemented jwt auth;
feat - Added 2 new custom exception;
feat - Added new methods in user service, controller;
feat - Added auth service and controller;
fix - Fixed create, update methods in user service;

Commit:
feat - Replace jwt key in user secret;

Commit:
fix - Fixed authorize in user controller;

Vladislawv commented on MyBlog.API/Controllers/AuthController.cs:
```diff
@@ -0,0 +1,57 @@
+[Route("api/auth")]
+[ApiController]
+public class AuthController : ControllerBase
+{
```
missed Produces tag

Vladislawv commented on MyBlog.API/Program.cs:
```diff
@@ -20,19 +25,61 @@
+            ValidAudience = builder.Configuration["Jwt:Audience"],
+            IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(builder.Configuration["Jwt:Key"]!))
+            
+        };
```
empty line

Vladislawv commented on MyBlog.API/Program.cs:
```diff
@@ -20,19 +25,61 @@
     config.IncludeXmlComments(xmlPath);
+    
+    config.AddSecurityDefinition("Bearer", new OpenApiSecurityScheme()
+    {
```
unnecessary (), if you call new ObjectInstance {}, you don`t need ().

Vladislawv commented on MyBlog.API/appsettings.json:
```diff
@@ -8,5 +8,10 @@
+    "Key": "Stores in UserSecrets",
+    "Issuer": "Nekitos",
+    "Audience": "AnyServer"
+  },
```
set your values to adecuate as possible

Vladislawv commented on MyBlog.Services/Areas/Auth/AuthService.cs:
```diff
@@ -0,0 +1,73 @@
+        var user = AuthenticateAsync(userLogin);
+
+        var token = GenerateJwtTokenAsync(await user);
+        
```
what is await user ?
it`s better to call await when you call async method

Vladislawv commented on MyBlog.Services/Areas/Auth/AuthService.cs:
```diff
@@ -0,0 +1,73 @@
+        var token = GenerateJwtTokenAsync(await user);
+        
+        return await token;
+    }
```
also call await only when you call method
it will be better to read and test

Vladislawv commented on MyBlog.Services/Areas/Auth/AuthService.cs:
```diff
@@ -0,0 +1,73 @@
+        };
+
+        var token = new JwtSecurityToken(_config["Jwt:Issuer"],
+            _config["Jwt:Audience"],
```
if you can`t write all method params by one line, split all params with another line.
In your case, pretty good, but you should also move first param to new line

Vladislawv commented on MyBlog.Services/Areas/Auth/AuthService.cs:
```diff
@@ -0,0 +1,73 @@
+    {
+        var user = await _userService.GetByEmailAsync(userLogin.Email);
+
+        if (userLogin.Password != user.Password)
```
also you can remove whitespace on the 41 line

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -37,8 +37,21 @@ public async Task<UserDto> GetByIdAsync(int id)
         return userDto;
     }
 
+    public async Task<User> GetByEmailAsync(string email)
+    {
+        var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == email)
+                   ?? throw new NotFoundException($"User with Email: {email} is not fount");
+        
+        return user;
```
rename all lamda variables to understandable, for example : u => user

is not fount ))))

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -37,8 +37,21 @@ public async Task<UserDto> GetByIdAsync(int id)
+        {
+            throw new ExistsException($"User with Email: {userInput.Email} or phone: {userInput.Phone} exists");
+        }
+        
```
it is good, that you improve new custom Exception, but i think BadRequest is better, because you write API, and imagine, a lot of people will be send requests to your api, and they all know about BadRequest, it is standart, but what about ExistsException ?They are new about that
also, you can remove {userInput.Email} and {userInput.Phone} - user know, if it`s wrong, that will be email or password

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -52,6 +65,11 @@ public async Task<int> UpdateByIdAsync(int id, UserDtoInput userInput)
                    ?? throw new NotFoundException($"User with Id: {id} is not found");
 
+        if (CheckForFreeAsync(id, userInput).Result == false)
+        {
```
it`s better to call methods and save it in variables, and than
call if statement with var

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -68,4 +86,43 @@ public async Task DeleteByIdAsync(int id)
+    private async Task<bool> CheckForFreeAsync(int id, UserDtoInput userInput)
+    {
+        var user = _context.Users.FirstOrDefault(u => u.Id == id)!;
+
```
remove !

Vladislawv replied:
also, you use async modificator in the method, why you do not use FirstOrDefaultAsync overload, and do not call await statement ?

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -68,4 +86,43 @@ public async Task DeleteByIdAsync(int id)
+
+        var repeatUserEmail = user.Email == userInput.Email;
+        var repeatUserPhone = user.Phone == userInput.Phone;
+
```
if you create bool var, it`s comfortable to read var like isUserEmailExist or isUserPhoneExist

Vladislawv replied:
also, i did not see Update method implementation with new feature

Commit:
fix - Fixed customs exceptions;
fix - Fixed dry code in user service;
fix - Changed return some methods in user service;
fix - Fixed unless async in generate jwt token.

Commit:
fix - Changed old version fluent validation on new.

Commit:
fix - Fixed code review remarks.

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -57,15 +55,16 @@ public async Task<int> CreateAsync(UserDtoInput userInput)
 
-        if (CheckForFreeAsync(id, userInput).Result == false)
+        if (await CheckEmailAndPhoneForFreeAsync(id, userInput) == false)
         {
```
you are in the if case, so if you use this bool expression, false condition replace to the "!"

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -75,21 +74,38 @@ public async Task<int> UpdateByIdAsync(int id, UserDtoInput userInput)
+        var user = await _context.Users.FirstOrDefaultAsync(u => u.Id == id) 
+                   ?? throw new NotFoundException($"User with Id: {id} is not found");
+
+        return user;
```
lamda variables naming

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -75,21 +74,38 @@ public async Task<int> UpdateByIdAsync(int id, UserDtoInput userInput)
+        var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == email)
+                   ?? throw new NotFoundException($"User with Email: {email} is not fount");
+
+        return user;
```
is not found, not fount
lamba variables

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -75,21 +74,38 @@ public async Task<int> UpdateByIdAsync(int id, UserDtoInput userInput)
     
-    private async Task<bool> CheckForFreeAsync(int id, UserDtoInput userInput)
+    private async Task<bool> CheckEmailAndPhoneForFreeAsync(int id, UserDtoInput userInput)
     {
```
it`s better to rename method CheckIfEmailAndPhoneAreFreeAsync

Vladislawv replied:
it`s better to rename method CheckIfEmailAndPhoneAreFreeAsync

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -75,21 +74,38 @@ public async Task<int> UpdateByIdAsync(int id, UserDtoInput userInput)
     {
-        var user = _context.Users.FirstOrDefault(u => u.Id == id)!;
+        var user = await _context.Users.FirstOrDefaultAsync(u => u.Id == id);
 
```
lambda vars

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -57,15 +55,16 @@ public async Task<int> CreateAsync(UserDtoInput userInput)
         await _context.Users.AddAsync(user);
```
lamba vars !!!!

Vladislawv commented on MyBlog.Services/Exception/BadRequestException.cs:
```diff
@@ -4,7 +4,7 @@ namespace MyBlog.Service.Exception;
 {
-    private const HttpStatusCode StatusCode = HttpStatusCode.BadRequest;
+    private new const HttpStatusCode StatusCode = HttpStatusCode.BadRequest;
     
```
why do you use new here ?

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -64,9 +64,10 @@ public async Task<UserDto> UpdateByIdAsync(int id, UserDtoInput userInput)
-        if (await CheckEmailAndPhoneForFreeAsync(id, userInput) == false)
+        var freeEmailAndPhone = await CheckEmailAndPhoneForFreeAsync(id, userInput);
+        if (freeEmailAndPhone == false)
         {
```
Rule for bool:
start name it with "is", so bool variable will be answer isEmailAndPhoneFree ?

Vladislawv replied:
if (!isEmailAndPhoneFree)
{
}

Vladislawv commented on MyBlog.Services/Areas/Users/UserService.cs:
```diff
@@ -89,26 +90,26 @@ public async Task DeleteByIdAsync(int id)
-        var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == email)
+        var user = await _context.Users.FirstOrDefaultAsync(user => user.Email == email)
                    ?? throw new NotFoundException($"User with Email: {email} is not fount");
 
```
not found

Vladislawv replied:
rename it in solution

Commit:
fix - Fixed code review remarks.
